File: benchmark/src/run_mean_prediction.R
library(SingleCellExperiment)
library(tidyverse)
Sys.setenv("BASILISK_EXTERNAL_CONDA"="/g/easybuild/x86_64/Rocky/8/haswell/software/Miniforge3/24.1.2-0")

pa <- argparser::arg_parser("Run model that always just predicts the mean across conditions")
pa <- argparser::add_argument(pa, "--dataset_name", type = "character", help = "The name of the dataset") 
pa <- argparser::add_argument(pa, "--test_train_config_id", type = "character", help = "The ID of the test/train/holdout run") 
pa <- argparser::add_argument(pa, "--seed", type = "integer", default = 1, nargs = 1, help = "The seed")

pa <- argparser::add_argument(pa, "--working_dir", type = "character", help = "The directory that contains the params, results, scripts etc.")
pa <- argparser::add_argument(pa, "--result_id", type = "character", help = "The result_id")
pa <- argparser::parse_args(pa)
# pa <- argparser::parse_args(pa, argv = r"(
#                             --dataset_name adamson
#                             --test_train_config_id ede703be5dd78-42650ea02bb4f
#                             --working_dir /scratch/ahlmanne/perturbation_prediction_benchmark
# )" |> stringr::str_trim() |> stringr::str_split("\\s+"))

print(pa)
set.seed(pa$seed)

out_dir <- file.path(pa$working_dir, "results/", pa$result_id)
# ---------------------------------------

# Load data
folder <- "data/gears_pert_data"
sce <- zellkonverter::readH5AD(file.path(folder, pa$dataset_name, "perturb_processed.h5ad"))
set2condition <- rjson::fromJSON(file = file.path(pa$working_dir, "results", pa$test_train_config_id))
if(! "ctrl" %in% set2condition$train){
  set2condition$train <- c(set2condition$train, "ctrl")
}

# Only keep valid condtions
sce <- sce[,sce$condition %in% unlist(set2condition)]

# Clean up the colData(sce) a bit
sce$condition <- droplevels(sce$condition)
sce$clean_condition <- stringr::str_remove(sce$condition, "\\+ctrl")
training_df <- tibble(training = names(set2condition), condition = set2condition) %>%
  unnest(condition)
colData(sce) <- colData(sce) %>%
  as_tibble() %>%
  tidylog::left_join(training_df, by = "condition") %>%
  DataFrame()

gene_names <- rowData(sce)[["gene_name"]]
rownames(sce) <- gene_names


# Pseudobulk everything
psce <- glmGamPoi::pseudobulk(sce, group_by = vars(condition, clean_condition, training))

# Calculate mean across training
mean_across_training <- MatrixGenerics::rowMeans2(assay(psce, "X")[,psce$training == "train",drop=FALSE])

# Bring into right format
pred <- replicate(nrow(psce), unname(mean_across_training), simplify = FALSE)
names(pred) <- psce$clean_condition

# Store output
tmp_out_dir <- file.path(tempdir(), "prediction_storage")
dir.create(tmp_out_dir)
write_lines(rjson::toJSON(pred), file.path(tmp_out_dir, "all_predictions.json"))
write_lines(rjson::toJSON(rownames(psce)), file.path(tmp_out_dir, "gene_names.json"))
file.rename(tmp_out_dir,out_dir)


#### Session Info
sessionInfo()

File: benchmark/src/run_mean_prediction.py
import argparse
import json
import shutil
import tempfile
from pathlib import Path

import numpy as np
import anndata as ad
import session_info

# --- Args -------------------------------------------------------------------

parser = argparse.ArgumentParser(description="Run model that always predicts the mean across conditions")
parser.add_argument("--dataset_name", required=True)
parser.add_argument("--test_train_config_id", required=True)
parser.add_argument("--seed", default=1, type=int)
parser.add_argument("--working_dir", required=True)
parser.add_argument("--result_id", required=True)
args = parser.parse_args()
print(args)

np.random.seed(args.seed)

out_dir = Path(args.working_dir) / "results" / args.result_id
# out_dir.parent.mkdir(parents=True, exist_ok=True) # No need to create out_dir.parent since it should already exist from the split step

# --- Load data --------------------------------------------------------------

adata = ad.read_h5ad(f"data/gears_pert_data/{args.dataset_name}/perturb_processed.h5ad")

with open(Path(args.working_dir) / "results" / args.test_train_config_id) as f:
    set2condition = json.load(f)

if "ctrl" not in set2condition["train"]:
    set2condition["train"].append("ctrl")

all_conditions = {c for conds in set2condition.values() for c in conds}
adata = adata[adata.obs["condition"].isin(all_conditions)].copy()
adata.obs["condition"] = adata.obs["condition"].astype("category").cat.remove_unused_categories()
adata.obs["clean_condition"] = adata.obs["condition"].str.replace(r"\+ctrl$", "", regex=True)

training_map = {cond: split for split, conds in set2condition.items() for cond in conds}
adata.obs["training"] = adata.obs["condition"].map(training_map)

gene_names = adata.var["gene_name"].values.tolist()
adata.var_names = gene_names

# --- Pseudobulk -------------------------------------------------------------

X = adata.X if isinstance(adata.X, np.ndarray) else adata.X.toarray()  # (cells, genes)

groups = (adata.obs[["condition", "clean_condition", "training"]]
          .drop_duplicates().reset_index(drop=True))

pb_X = np.zeros((len(groups), X.shape[1]))  # (groups, genes)
for i, row in groups.iterrows():
    mask = (adata.obs["condition"] == row["condition"]).values
    pb_X[i] = X[mask].mean(axis=0)

# --- Mean across training conditions ----------------------------------------

train_mask = (groups["training"] == "train").values
mean_across_training = pb_X[train_mask].mean(axis=0)  # (genes,)

# Assign that same vector to every condition
pred = {cond: mean_across_training.tolist()
        for cond in groups["clean_condition"]}

# --- Save output ------------------------------------------------------------

tmp_out_dir = Path(tempfile.mkdtemp())
(tmp_out_dir / "all_predictions.json").write_text(json.dumps(pred), encoding="utf-8")
(tmp_out_dir / "gene_names.json").write_text(json.dumps(gene_names), encoding="utf-8")

if out_dir.exists():
    shutil.rmtree(out_dir)
shutil.move(str(tmp_out_dir), out_dir)

session_info.show()
print("Python done")